Rationale.md

```diff
@@ -0,0 +1,230 @@
+map the underlying OS's capabilities to the WebAssembly application, as well as
+to communicate their own implementation details in a useful manner to the
+developer.
+
+
+## Control Flow
+
+See [#299](https://github.com/WebAssembly/design/pull/299).
+
+
+## Locals
+
+C/C++ makes it possible to take the address of a function's local values and
+pass this pointer to callees or to other threads. Since WebAssembly's local
+variables are outside the address space, C/C++ compilers implement address-taken
+variables by creating a separate stack data structure within linear memory. This
+stack is sometimes called the "aliased" stack, since it is used for variables
+which may be pointed to by pointers.
+
+This prevents WebAssembly from performing clever optimizations on the stack and
+liveness of such variables, but this loss isn't expected to be
+consequential. Common C compiler optimizations such as LLVM's global value
+numbering effectively split address-taken variables into parts, shrinking the
+range where they actually need to have their address taken, and creating new
+ranges where they can be allocated as local variables.
+
+Conversely, non-address taken values which are usually on the stack are instead
+represented as locals inside functions. This effectively means that WebAssembly
+has an infinite set of registers, and can choose to spill values as it sees fit
+in a manner unobservable to the hosted code. This implies that there's a
+separate stack, unaddressable from hosted code, which is also used to spill
+return values. This allows strong security properties to be enforced, but does
+mean that two stacks are maintained (one by the VM, the other by the compiler
+which targets WebAssembly) which can lead to some inefficiencies.
+
+Local variables are not in Static Single Assignment (SSA) form, meaning that
+multiple incoming SSA values which have separate liveness can "share" the
+storage represented by a local through the `set_local` operation. From an SSA
+perspective, this means that multiple independent values can share a local
+variable in WebAssembly, which is effectively a kind of pre-coloring that clever
+producers can use to pre-color variables and give hints to a WebAssembly VM's
+register allocation algorithms, offloading some of the optimization work from
+the WebAssembly VM.
+
+
+## Variable-Length Argument Lists
+
+C and C++ compilers are expected to implement variable-length argument lists by
+storing arguments in a buffer in linear memory and passing a pointer to the
+buffer. This greatly simplifies WebAssembly VM implementations by punting this
+ABI consideration to the front-end compiler. It does negatively impact
+performance, but variable-length calls are already somewhat slow.
+
+
+## Multiple Return Values
+
+TODO
+
+
+## Indirect Calls
+
+The exact semantics of indirect function calls, function pointers, and what
+happens when calling the wrong function, are still being discussed.
```

Comment:
This is OK to land (since it was just yanked from elsewhere), but it will need to be revised because it's outdated.